tree/treeplayer/test/gh16804.cxx

```diff
@@ -0,0 +1,388 @@
+
+   int index = -1, value = -1, mainBranch = -1, a = -1, b = -1;
+   auto mainBranchRet = mainTree->SetBranchAddress("mainBranch", &mainBranch);
+   auto indexRet = mainTree->SetBranchAddress("index", &index);
+   auto valueRet = mainTree->SetBranchAddress("value", &value);
+   auto aRet = mainTree->SetBranchAddress("a", &a);
+   auto bRet = mainTree->SetBranchAddress("b", &b);
+   EXPECT_GE(mainBranchRet, 0);
+   EXPECT_GE(indexRet, 0);
+   EXPECT_GE(valueRet, 0);
+   EXPECT_GE(aRet, 0);
+   EXPECT_GE(bRet, 0);
+
+   std::vector<int> expectedMainBranch(20);
+   std::vector<int> expectedIndex(20);
+   std::vector<int> expectedValue(20);
+   std::vector<int> expectedA(20);
+   std::vector<int> expectedB(20);
+   std::iota(expectedMainBranch.begin(), expectedMainBranch.end(), 0);
+   std::reverse(expectedMainBranch.begin(), expectedMainBranch.end());
+   std::iota(expectedIndex.begin(), expectedIndex.end(), 0);
+   std::iota(expectedValue.begin(), expectedValue.end(), 100);
+   std::iota(expectedA.begin(), expectedA.end(), 200);
+   std::iota(expectedB.begin(), expectedB.end(), 300);
+   auto nEntries = mainTree->GetEntriesFast();
+
+   // FIXME: The following diagnostics should not happen. This is the situation:
+   // We have main tree with branch "mainBranch", friend tree 1 with branches
+   // "index","value", friend tree 2 with branches "a","b". In the previous calls
+   // to SetBranchAddress, we registered *all* branches "index","value","a","b"
+   // as TChainElements of the two friend TChains. This has to be like this, since
+   // a priori we cannot know which branch belongs to which TChain as the trees
+   // haven't been loaded yet (and SetBranchAddress does not trigger loading the first
+   // tree of a TChain). What ends up happening is that "index" and "value" are
+   // correctly paired with friend 1, as well as "a" and "b" with friend 2. But,
+   // crucially, friend 1 still has TChainElements for "a" and "b" and viceversa
+   // for friend 2. So, when calling LoadTree on friend 1, SetBranchStatus is
+   // called for all TChainElements, leading to friend 1 looking for branches
+   // "a" and "b" and not being able to find them. Again, the converse happens
+   // for friend 2. This happens at every file boundary, so the errors are printed
+   // 2 times for friend 1 and 4 times for friend 2. It is yet unclear how this
+   // could be prevented. Note that the values are properly found and loaded for
+   // all branches of all friends and in fact the EXPECT_EQ calls in the
+   // following for loop all pass.
+   ROOT::TestSupport::CheckDiagsRAII diagRAII;
+   diagRAII.requiredDiag(kError, "TTree::SetBranchStatus", "unknown branch -> a");
+   diagRAII.requiredDiag(kError, "TTree::SetBranchStatus", "unknown branch -> b");
+   diagRAII.requiredDiag(kError, "TTree::SetBranchStatus", "unknown branch -> index");
+   diagRAII.requiredDiag(kError, "TTree::SetBranchStatus", "unknown branch -> value");
```

Comment:
See explanation on current issue with spurious error messages

Reply:
Hmm this is indeed annoying. From the description, I think we'll also want a test where both friend chains share a branch. It should always load values from the first chain.

Reply:
Indeed by only using `SetBranchAddress` the branch read is the one from the first TTree who has the branch "x", any other TTree who might have the same branch name is disregarded.

Reply:
But does this work with chains? Since the full branch list isn't known before the first GetEntry...

Reply:
Just tested, works with TChain the same way

Reply:
Aren't those message directly linked to https://github.com/root-project/root/pull/18871#discussion_r2109406331 (Even-though the propose solution is too simplistic).

Reply:
Or we force the chain to load, like `TChain` already does in some cases: `FindBranch()`, `FindLeaf()`, `GetAlias()`, etc:
https://github.com/root-project/root/blob/40d2a622d3fe1b876c8c31a04edd37c25423a6ef/tree/tree/src/TChain.cxx#L930-L933

Reply:
This is essentially what we ended up with.